### PR TITLE
ansible@2.9 2.9.18

### PR DESCRIPTION
**Pull request by laserscout**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by laserscout** [comment]
I opened a draft PR as I would like to also fix the `pypi_formula_mappings.json` for this formula

**Comment by laserscout** [comment]
I think that I'm done with my clean up.

I restructured the formula to a) make it clear which packages are installed and for what reason, avoiding dragging along old dependencies; and b) make it easier for anyone to update the formula and its dependencies in the future.

I've left my dirty-ish commits as they were to allow the reviewer to more easily look at the changes and their reasoning.

Here is a list of all changes:
* bump ansible to 2.9.18
* remove python2 packages
  * boto, we already use boto3
  * use pyshpere3 instead of pysphere
* remove backport packages, ansible@2.9 is build on python 3.9
  * backports.ssl_match_hostname
  * contextlib2
  * funcsigs
  * ipaddress
  * monotonic
  * unicodecsv
* remove seemingly unused packages
  * babel
  * deprecation
  * docutils
  * subprocess32
* remove asn1crypto as it's not needed since cyptography 2.7
* remove dictdiffer as it's not needed since openshift<0.11.0
* Remove the permission fix for prettytable 0.7.2 as we're using prettytable 2.1.0
* Found resource dependency heads for existing packages and added them in `pypi_formula_mappings.json`
* Update resources
* Restructure formula and added documentation.

The formula was updated by hand and through `brew update-python-resources` yielding the same resulting formula.

File: Formula/ansible@2.9.rb [hunk 1]
```diff
@@ -34,135 +33,71 @@ class AnsibleAT29 < Formula
...
   uses_from_macos "libxslt"
 
-  # For instructions on how to update the resource blocks, see:
-  # https://github.com/Homebrew/homebrew-core/pull/50504#issuecomment-592743222
-
-  # Collect requirements from:
+  # The first, long block of resources can be semi-automatically updated by two similar ways
+  # Either run `brew update-python-resources ansible@2.9 --print-only` and replace the
+  # resource block manually.
+  # Or comment out the secord brock with the manual resource and run
```
Comment (carlocab) [inline Formula/ansible@2.9.rb:39]:
```suggestion
  # Or comment out the second block with the manual resource and run
```

**Review by carlocab — COMMENTED** [review]
Thanks very much for all your work on this, @laserscout. I only have a small, not-very-substantive, suggestion.

**Comment by laserscout** [comment]
> Thanks very much for all your work on this, @laserscout. I only have a small, not-very-substantive, suggestion.

Thank you for the review and the typo fix.

Squash time then?

**Comment by carlocab** [comment]
Let's wait to hear from @Rylan12 and/or @samford on this.

File: pypi_formula_mappings.json [hunk 2]
```diff
@@ -7,6 +7,14 @@
       "extra_packages": ["ansible-base"]
   },
   "ansible@2.8": false,
+  "ansible@2.9": {
+    "extra_packages": [
+      "apache-libcloud", "boto3", "dnspython", "docker-py", "junos-eznc", "jxmlease", "kerberos", "ntc-templates",
+      "openshift", "passlib", "pexpect", "pysphere3", "python-consul", "python-keyczar", "python-neutronclient",
+      "pywinrm","requests-credssp","shade", "zabbix-api"
```
Comment (Rylan12) [inline pypi_formula_mappings.json:14]:
```suggestion
      "pywinrm", "requests-credssp", "shade", "zabbix-api"
```

**Review by Rylan12 — COMMENTED** [review]
This is great!

Other than the one (totally minor) syntax thing I pointed out, I'd be fine merging this.

I am curious, though, as to why there are so many PyPI packages that aren't included by default. Are we sure these are all being used? Most PyPI formulae don't seem to need to have all of these added. Are they in their most-grouped form? Meaning, are all of these really needed, or are some of them just dependencies of others that we don't need to explicitly list?

Side note: I noticed that the old file has [a comment](https://github.com/Homebrew/homebrew-core/blob/7136ea98e53d86ef24b5698d09c409cde47080b3/Formula/ansible%402.9.rb#L76) blocking off resources for `requests[security]`, but I don't see `requests[security]` in the `extra_packages`. Clearly, this isn't much of an issue because the [`requests[security]` test](https://github.com/Homebrew/homebrew-core/blob/f46e7ca7871888448792eb0118954ee2f18d8616/Formula/ansible%402.9.rb#L630-L632) still passes. I just want to make sure that this isn't an oversight.

**Comment by laserscout** [comment]
@Rylan12 

> why there are so many PyPI packages that aren't included by default.

Unfortunately, the path Ansible chose regarding dependencies is a little flaky. Installing ansible via pip will only install the bare minimum required packages for it to run and forgo dependencies even for the by default included plugins and modules.

Let first take `apache-libcloud` as an example. It was added on [149703aa90 (ansible: add apache-libcloud to support GCE., 2016-02-01)](https://github.com/Homebrew/homebrew-core/commit/149703aa90c7c373284b0873394fa06a74334ff9) to, as the message says, add GCE support. On the one hand, [current high level documentation](https://docs.ansible.com/ansible/2.9/scenario_guides/guide_gce.html) **does not** mention `apache-libcloud` as a dependency of the GCE family of modules. On the other, searching for the string "libcloud" in ansible/stable-2.9 returns several matches some of which under `./lib/ansible/modules/cloud/google`. Searching for individual module documentation based of these findings ie: [gce_mig](https://docs.ansible.com/ansible/2.9/modules/gce_mig_module.html#requirements), [gce_lb](https://docs.ansible.com/ansible/2.9/modules/gce_lb_module.html#requirements), [gce_net](https://docs.ansible.com/ansible/2.9/modules/gce_net_module.html#requirement), etc... mentions that `apache-libcloud` is required can be found.

As as second, much simpler example, the [dig lookup plugin](https://docs.ansible.com/ansible/2.9/plugins/lookup/dig.html#requirements), which is a build in feature, requires `dnspython` but isn't a dependency of the `ansible` package.

> Are we sure these are all being used?

I already removed some packages which I felt confident aren't used any more, but some more might need to be removed. You saw how in depth analysis was required to understand the usage of `apache-libcloud`...

> Are they in their most-grouped form?

As mentioned in my previous long comment, I did my best using `pipdeptree` to get the dependency heads from all of these resources.

> Side note: I noticed that the old file has [a comment](https://github.com/Homebrew/homebrew-core/blob/7136ea98e53d86ef24b5698d09c409cde47080b3/Formula/ansible%402.9.rb#L76) blocking off resources for `requests[security]`, but I don't see `requests[security]` in the `extra_packages`. Clearly, this isn't much of an issue because the [`requests[security]` test](https://github.com/Homebrew/homebrew-core/blob/f46e7ca7871888448792eb0118954ee2f18d8616/Formula/ansible%402.9.rb#L630-L632) still passes. I just want to make sure that this isn't an oversight.

`cryptography` is now a dependency of `ansible` and `requests`, `idna`, and `pyOpenSSL` are dependencies for several other installed packages so I didn't feel the need to explicitly add them. The test of course remains to ensure future updates still work.

**Comment by carlocab** [comment]
I think you're good to squash your commits together now.

**Review by Rylan12 — APPROVED** [review]
Thanks for the thorough explanation, @laserscout. I'm 👍 on merging this. I'd rather not accidentally remove features that we've already had to manually add so I think it's good to err on the side of keeping those dependencies around.

In terms of the `requests[security]`, I think it's not a bad idea to include it in case any of those packages that you mentioned stop being dependencies of the other packages in the future. Because it's not explicitly dependent on `requests[security]`, it's possible for things to be missed. Yes, though, the test should catch it.

Either way, LGTM. I'll go ahead and squash for you so we can get this shipped as it's a version bump rather than simply a build change.

**Comment by carlocab** [comment]
Thanks, @laserscout. Any chance you can have a look at the `ansible` formula too?

**Comment by laserscout** [comment]
@carlocab I was thinking about it, but I think that the `Ansible` formula needs further clean up and some community discussions.

My reason for thinking this again has to do with package structure chosen by Ansible. They [restructured](https://www.ansible.com/blog/thoughts-on-restructuring-the-ansible-project) their project and split the ansible core, providing the `ansible` commands, and the rest of the modules, plugins, etc... So now, [after the restructure](https://www.ansible.com/blog/ansible-3.0.0-qa) the latest `ansible-3.1.0` package installs a collection of modules, plugins, etc... which depends on `ansible-base-2.10.7`. So you are actually installing ansible 2.10

```
> pip3 install ansible
Collecting ansible
  Using cached ansible-3.1.0.tar.gz (31.1 MB)
Collecting ansible-base<2.11,>=2.10.5
  Downloading ansible-base-2.10.7.tar.gz (5.7 MB)
     |████████████████████████████████| 5.7 MB 2.9 MB/s 
Collecting jinja2
  Using cached Jinja2-2.11.3-py2.py3-none-any.whl (125 kB)
Collecting PyYAML
  Using cached PyYAML-5.4.1-cp39-cp39-macosx_10_9_x86_64.whl (259 kB)
Collecting cryptography
  Using cached cryptography-3.4.6-cp36-abi3-macosx_10_10_x86_64.whl (2.0 MB)
Collecting packaging
  Using cached packaging-20.9-py2.py3-none-any.whl (40 kB)
Collecting cffi>=1.12
  Using cached cffi-1.14.5-cp39-cp39-macosx_10_9_x86_64.whl (177 kB)
Collecting pycparser
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Collecting MarkupSafe>=0.23
  Using cached MarkupSafe-1.1.1-cp39-cp39-macosx_10_9_x86_64.whl (16 kB)
Collecting pyparsing>=2.0.2
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Using legacy 'setup.py install' for ansible, since package 'wheel' is not installed.
Using legacy 'setup.py install' for ansible-base, since package 'wheel' is not installed.
Installing collected packages: pycparser, pyparsing, MarkupSafe, cffi, PyYAML, packaging, jinja2, cryptography, ansible-base, ansible
    Running setup.py install for ansible-base ... done
    Running setup.py install for ansible ... done
Successfully installed MarkupSafe-1.1.1 PyYAML-5.4.1 ansible-3.1.0 ansible-base-2.10.7 cffi-1.14.5 cryptography-3.4.6 jinja2-2.11.3 packaging-20.9 pycparser-2.20 pyparsing-2.4.7
> ansible --version | head -n1
ansible 2.10.7
```

This sort of restructuring affects how brew would version its ansible formulas. Maybe (if it's possible) an Ansible-base package should exist in brew as well which the ansible formula would depend on?

Where could we have this sort of discussion 🤔

**Comment by carlocab** [comment]
We can have a discussion in a draft PR.

Not sure if it's necessary to have an `ansible-base` formula. It might be better to just install `ansible-base` into the global environment (instead of the formula-specific venv) in the `ansible` formula. Then, formulae that need `ansible-base` (e.g. `ansible-lint`) can just depend on `ansible`.

To install `ansible-base` into the global environment, we can do something like this upon staging the `ansible-base` resource:

https://github.com/Homebrew/homebrew-core/blob/fbb7ef4b2e8c88115b1f7dd43ae1cd93130c1dc2/Formula/hashpump.rb#L37

We'd have to exclude `ansible-base` from the resources that are installed into the `ansible`-specific venv, but that's not too hard to do. (See, for example, formulae that use the resource `Pillow`.)

**Pull request closed without merging**